### PR TITLE
Allow sales numbers to be more dynamic

### PR DESCRIPTION
**Pull request by Cruiser13**
| Q             | A
| ------------- | ---
| Bug fix?      | no
| New feature?  | yes
| BC breaks?    | no
| Deprecations? | no
| Fixed tickets | 

This would give us the possibility to use numbers like:
O2019-12-12345678
by prefix config:
O{year}-{month}-

Similar to the shopware feature: https://docs.shopware.com/en/shopware-6-en/settings/Numberranges

The current implementation in this pull request is just a base for discussing this proposal.

### CODE REVIEW

**Comment by dpfaffenbauer** [comment]
Still not dynamic enough. Can you change it to use at least the expression engine?

**Comment by Cruiser13** [comment]
@dpfaffenbauer I agree, this is just a base for discussing this feature. We should consider more options than just using the few date entries I put in there. Depending on the complexity we would like to include we should go for different solutions.

**Comment by dpfaffenbauer** [comment]
Expression Engine... that is extendable

**Comment by Cruiser13** [comment]
Changed accordingly. Not sure if we should add other contexts though (as coreshop does with the ExpressionInterpreter).

**Comment by dpfaffenbauer** [comment]
... please use proper dependency injection

File: src/CoreShop/Component/Core/Order/NumberGenerator/SaleNumberGenerator.php [hunk 1]
```diff
@@ -72,7 +73,14 @@ public function generate(ResourceInterface $model)
         }
 
         if ($store instanceof StoreInterface) {
-            return sprintf('%s%s%s', $this->configurationService->getForStore($this->prefixConfigurationKey, $store), $this->numberGenerator->generate($model), $this->configurationService->getForStore($this->suffixConfigurationKey, $store));
+            $prefix = $this->configurationService->getForStore($this->prefixConfigurationKey, $store);
+            $suffix = $this->configurationService->getForStore($this->suffixConfigurationKey, $store);
+            
+            $expr = new ExpressionLanguage();
```
Comment (dpfaffenbauer) [inline src/CoreShop/Component/Core/Order/NumberGenerator/SaleNumberGenerator.php]: inject via DI

**Comment by Cruiser13** [comment]
Sorry @dpfaffenbauer I missed your comments. Thanks for the modification!

**Comment by dpfaffenbauer** [comment]
Actually, thinking about it, this is a BC break.

Since currently we have default settings like prefix "O" for Order. But, if you pass "O" into the Expression Engine, it throws an exception, cause "O" is not a valid expression, should be "\"O\"" to be valid.

If you need that feature, you'd have to wait for 3.0, or decorate the Generator in your project and add that

**Comment by Cruiser13** [comment]
@dpfaffenbauer as 3.0 is now prepared this might be a good include for 3.0?

**Comment by dpfaffenbauer** [comment]
Yes.

**Pull request closed without merging**